include/internal/quic_cc.h

```diff
@@ -14,146 +14,205 @@
+                            const OSSL_PARAM *params);
 
     /*
-     * Get current number of packets exempted from CC.
-     * Returns negative value on error, the number otherwise.
+     * (Re)bind output (diagnostic) information.
+     *
+     * params is an array of OSSL_PARAM structures used to output values. The
+     * storage locations associated with each parameter are stored internally
+     * and updated whenever the state of the congestion controller is updated;
+     * thus, the storage locations associated with the OSSL_PARAMs passed in the
+     * call to this function must remain valid until the congestion controller
+     * is freed or those parameters are unbound. A given parameter name may be
+     * bound to only one location at a time. The params structures themselves
+     * do not need to remain allocated after this call returns.
+     *
+     * Returns 1 on success and 0 on failure.
      */
-    int (*get_exemption)(OSSL_CC_DATA *ccdata);
+    int (*bind_diagnostics)(OSSL_CC_DATA *ccdata,
+                            OSSL_PARAM *params);
 
     /*
-     * Returns 1 if sending is allowed, 0 otherwise.
+     * Unbind diagnostic information. The parameters with the given names are
+     * unbound, cancelling the effects of a previous call to bind_diagnostic().
+     * params is an array of OSSL_PARAMs. The values of the parameters are
+     * ignored. If a parameter is already unbound, there is no effect for that
+     * parameter but other parameters are still unbound.
+     *
+     * Returns 1 on success or 0 on failure.
      */
-    int (*can_send)(OSSL_CC_DATA *ccdata);
+    int (*unbind_diagnostics)(OSSL_CC_DATA *ccdata,
+                              OSSL_PARAM *params);
 
     /*
-     * Returns number of bytes allowed to be sent.
-     * |time_since_last_send| is time since last send operation
-     * in microseconds.
-     * |time_valid| is 1 if the |time_since_last_send| holds
-     * a meaningful value, 0 otherwise.
+     * Returns the amount of additional data (above and beyond the data
+     * currently in flight) which can be sent in bytes. Returns 0 if no more
+     * data can be sent at this time. The return value of this method
+     * can vary as time passes.
      */
-    uint64_t (*get_send_allowance)(OSSL_CC_DATA *ccdata,
-                                   OSSL_TIME time_since_last_send,
-                                   int time_valid);
+    uint64_t (*get_tx_allowance)(OSSL_CC_DATA *ccdata);
 
     /*
-     * Returns the maximum number of bytes allowed to be in flight.
+     * Returns the time at which the return value of get_tx_allowance might be
+     * higher than its current value. This is not a guarantee and spurious
+     * wakeups are allowed. Returns ossl_time_infinite() if there is no current
+     * wakeup deadline.
      */
-    uint64_t (*get_bytes_in_flight_max)(OSSL_CC_DATA *ccdata);
+    OSSL_TIME (*get_wakeup_deadline)(OSSL_CC_DATA *ccdata);
```

Comment:
When we make this provided, we won't be able to use OSSL_TIME. But IMO this is OK for now.